feature(authorization): Add creation access control

AbdulRahmanAlHamali opened the pull request:
**This implementation is still in draft state**. It achieves the following:
1. Add a configuration parameter called `restrictApplicationCreation`. If set to false (the default), this feature will be ignored.
2. Add a `CREATE` authorization.
3. Call the `ResourcePermissionProvider` immediately from `AuthorizeController`, because creation permissions are typically dynamic, and cannot be known before the application is submitted for creation. Thus, they cannot be held on the UserPermissions object.
4. Make `Front50ApplicationResourcePermission` source strip out any `CREATE` permissions provided on the source, because it doesn't make sense for users to add `CREATE` permissions on the resource they are creating, and it will basically work against any other authorization model we try to implement.
5.  Implementations of `ResourcePermissionProvider` and `ResourcePermissionSource` will be free to handle creation permissions the way they say fit.

Please let me know what you think. Let's try to release this in the Oct 28th release

AbdulRahmanAlHamali commented on fiat-web/src/main/java/com/netflix/spinnaker/fiat/controllers/AuthorizeController.java:
```diff
@@ -150,35 +147,54 @@ public AuthorizeController(
   }
 
```
I realize I should not modify the function below, and create a special one to handle `CREATE`, because the function below is name-based, while for `CREATE` I want to handle the whole object not just its name.
Nevertheless, I will keep this for now for you to look at, and will modify later.
It is also worth it to (in a different PR), make a breaking change that migrates all the usages to use the full resource instead of just the name.

AbdulRahmanAlHamali commented on fiat-web/src/main/java/com/netflix/spinnaker/fiat/controllers/AuthorizeController.java:
```diff
@@ -150,35 +147,54 @@ public AuthorizeController(
 
-    try {
-      switch (r) {
-        case ACCOUNT:
```
yup I actually never found any calls for it in other services

AbdulRahmanAlHamali commented on fiat-web/src/main/java/com/netflix/spinnaker/fiat/controllers/AuthorizeController.java:
```diff
@@ -128,6 +123,22 @@ public AuthorizeController(
         .orElseThrow(NotFoundException::new);
   }
 
```
I added these below since we didn't have them

cfieber replied:
I'd prefer to just leave these out - we don't seem to have a need for them anywhere yet

AbdulRahmanAlHamali commented:
@plumpy would you kindly provide feedback on this PR?

cfieber commented on fiat-api/src/main/java/com/netflix/spinnaker/fiat/shared/FiatPermissionEvaluator.java:
```diff
@@ -143,7 +147,51 @@ private static RetryHandler buildRetryHandler(
+      return true;
+    }
+
+    Resource r;
```
This is tricky, but I think we need to think through it a bit more

This code is actually running in-process in the fiat clients, and is annotated onto methods there. (For example in Front50 on a `getApplications` endpoint). In that case, the type is going to be that applications API model type, which isn't going to implement Resource.

Not sure I have a great suggestion immediately, thinking about it..

cfieber commented on fiat-web/src/main/java/com/netflix/spinnaker/fiat/controllers/AuthorizeController.java:
```diff
@@ -189,6 +200,64 @@ public void getUserAuthorization(
+  public void hasAuthorization(
+      @PathVariable String userId,
+      @PathVariable String authorization,
+      @RequestBody @Nonnull Resource resource,
```
Follow-on to the previous comment - we will need to see how we go from user domain object to Resource but this might need `@PathVariable ResourceType` and `@RequestBody Map<String, Object> resource` so we can handle the object conversion properly - currently the Resource hierarchy isn't set up for Jackson polymorphic deserialization.

cfieber commented on fiat-web/src/main/java/com/netflix/spinnaker/fiat/controllers/AuthorizeController.java:
```diff
@@ -189,6 +204,42 @@ public void getUserAuthorization(
+
+    Resource r = objectMapper.convertValue(resource, rt.modelClass);
+
+    // can easily implement options other than APPLICATION, but it is not currently needed.
```
+1 I'm happy with keeping the scope narrow to the existing use-case for now

cfieber commented on fiat-api/src/main/java/com/netflix/spinnaker/fiat/shared/FiatPermissionEvaluator.java:
```diff
@@ -146,6 +149,37 @@ public boolean hasPermission(
+      return AuthenticatedRequest.propagate(
+              () -> {
+                return retryHandler.retry(
+                    "get whether " + username + " can create resource" + resource,
```
```suggestion
                    "determine whether " + username + " can create resource " + resource,
```

cfieber commented on fiat-roles/src/main/java/com/netflix/spinnaker/fiat/providers/Front50ApplicationResourcePermissionSource.java:
```diff
@@ -53,6 +53,9 @@ public Permissions getPermissions(@Nonnull Application resource) {
       authorizations.put(Authorization.EXECUTE, authorizations.get(executeFallback));
     }
 
+    // CREATE permissions are not allowed on the resource level
```
```suggestion
    // CREATE permissions are not allowed on the resource level.
    //  At this point the resource already exists, so CREATE does
    //  not make sense.
```

cfieber commented on fiat-api/src/main/java/com/netflix/spinnaker/fiat/shared/FiatPermissionEvaluator.java:
```diff
@@ -146,6 +149,37 @@ public boolean hasPermission(
+                      } else if (response.getStatus() == HttpServletResponse.SC_NOT_FOUND) {
+                        return false;
+                      } else {
+                        throw new ServletException(
```
lets pick something other than ServletException here. I think IntegrationException with setRetryable(true) would be more appropriate (we may want to smarten up retry handler in the future, so there is a preference to use subclasses of SpinnakerException wherever possible)

cfieber replied:
should have spinnaker/kork#401 merged in to make that a litle less painful

cfieber replied:
will also not want to retry on BAD_REQUEST

cfieber replied:
see https://github.com/spinnaker/fiat/pull/485/files#r337840808

cfieber commented on fiat-web/src/main/java/com/netflix/spinnaker/fiat/controllers/AuthorizeController.java:
```diff
@@ -189,6 +188,42 @@ public void getUserAuthorization(
+    Resource r = objectMapper.convertValue(resource, rt.modelClass);
+
+    // can easily implement options other than APPLICATION, but it is not currently needed.
+    if (applicationResourcePermissionProvider
```
think we might need an isAdmin check here on user as well

cfieber commented on fiat-api/src/main/java/com/netflix/spinnaker/fiat/shared/FiatPermissionEvaluator.java:
```diff
@@ -146,6 +149,37 @@ public boolean hasPermission(
+                return retryHandler.retry(
+                    "determine whether " + username + " can create resource" + resource,
+                    () -> {
+                      Response response = fiatService.canCreate(username, resourceType, resource);
```
The response handling bit in here is incorrect.

On a 4xx response from the server, retrofit throws a RetrofitException

This should looks something like:

```suggestion
                      try {
                        fiatService.canCreate(username, resourceType, resource);
                        return true;
                      } catch (RetrofitError re) {
                        boolean shouldRetry = true;
                        if (re.getKind() == RetrofitError.Kind.HTTP) {
                          switch (HttpStatus.valueOf(re.getResponse().getStatus())) {
                            case NOT_FOUND:
                              return false;

                            case BAD_REQUEST:
                              shouldRetry = false;
                              break;
                          }
                        }
                        IntegrationException ie = new IntegrationException(re);
                        ie.setRetryable(shouldRetry);
                        throw ie;
                      }
```

AbdulRahmanAlHamali replied:
The retry handler doesn't look like it cares about the `retryable` property. Or is this for the future?

cfieber commented on fiat-api/src/main/java/com/netflix/spinnaker/fiat/shared/FiatPermissionEvaluator.java:
```diff
@@ -146,6 +149,37 @@ public boolean hasPermission(
+      return AuthenticatedRequest.propagate(
+              () -> {
+                return retryHandler.retry(
+                    "determine whether " + username + " can create resource" + resource,
```
```suggestion
                    "determine whether " + username + " can create resource " + resource,
```

cfieber commented on fiat-web/src/main/java/com/netflix/spinnaker/fiat/controllers/AuthorizeController.java:
```diff
@@ -189,6 +188,42 @@ public void getUserAuthorization(
+    if (rt != ResourceType.APPLICATION) {
+      response.sendError(
+          HttpServletResponse.SC_BAD_REQUEST,
+          "Resource type " + resourceType + "does not support creation");
```
```suggestion
          "Resource type " + resourceType + " does not support creation");
```

cfieber commented on fiat-api/src/main/java/com/netflix/spinnaker/fiat/shared/FiatService.java:
```diff
@@ -50,6 +50,18 @@ Response hasAuthorization(
+   * @return 200 if the user can create the specified resource, 404 otherwise
+   */
+  @POST("/authorize/{userId}/{resourceType}")
+  Response canCreate(
```
```suggestion
  void canCreate(
```
this can be void instead of Response - we only care about the Response if we got a RetrofitError and in that case we can get it from there

AbdulRahmanAlHamali commented:
> other than the RetrofitError handling and a couple minor spacing tweaks this looks good
> 
> I was able to build locally and integrate into front50 and add a ResourcePermissionSource that added a CREATE authorization to the application and see it properly enforced.
> 
> I am not proposing changing it here (lets revisit in the next PR that tries to add something supplying CREATE permissions for Applications), but I do suspect we will need to enhance the ResourcePermissionProvider/Source model to indicate that we are asking for Permissions for a new resource - in my contrived ResourcePermissionSource for testing I felt like it would be hard to write a legitimate ResourcePermissionSource to add CREATE and also support the case of not causing a resource to go from isRestricted=false to isRestricted=true.
> 
> But, let's dig into that in the next PR..

Alright, I have addressed your comments :)

Regarding distinguishing new resources, I am trying in this PR to get things done without distinguishing, but it might indeed be too hard. I'll start the next PR for the PrefixSource, which will make things clearer for us.
